--- selinux/file_contexts.old	2021-08-13 21:25:42.010662000 +0100
+++ selinux/file_contexts	2021-08-26 20:42:15.805971599 +0100
@@ -14,11 +14,16 @@
 
 /dev/akm8975					u:object_r:sensors_device:s0
 
+/dev/socket/dbus				u:object_r:dbus_daemon_socket:s0
+
 /system/bin/ks					u:object_r:ks_exec:s0
 /system/bin/qcks				u:object_r:qcks_exec:s0
 /system/bin/efsks				u:object_r:efsks_exec:s0
 /system/bin/qmuxd				u:object_r:qmuxd_exec:s0
 /system/bin/qmiproxy				u:object_r:qmiproxy_exec:s0
+/system/bin/dbus-daemon				u:object_r:dbus_daemon_exec:s0
+/system/bin/ofonod				u:object_r:ofono_exec:s0
+/system/bin/qmiserial2qmuxd			u:object_r:ofono_exec:s0
 
 /dev/ttyUSB0					u:object_r:serial_device:s0
 /dev/ttySAC0					u:object_r:serial_device:s0
