--- rootdir/ueventd.smdk4x12.rc.old	2021-08-13 21:25:42.010662000 +0100
+++ rootdir/ueventd.smdk4x12.rc	2021-08-15 17:01:57.373976760 +0100
@@ -75,6 +75,7 @@
 /dev/ttyUSB1		0660	system	radio
 /dev/ttyUSB2		0660	system	system
 /dev/hsic*		0666	system	radio
+/dev/cdc-wdm*		0660	system	radio
 /dev/usb/lp*		0660	system	usb
 /dev/ttyGS0		0660	system	system
 /sys/devices/virtual/input/input*	poll_delay	0660	system	radio
