--- gcc/common.opt.old	2023-03-25 14:21:04.348712898 +0000
+++ gcc/common.opt	2023-03-26 18:30:10.637768323 +0000
@@ -616,10 +616,6 @@
 Common RejectNegative Joined UInteger Warning
 -Wlarger-than=<number>	Warn if an object is larger than <number> bytes.
 
-Wnonnull-compare
-Var(warn_nonnull_compare) Warning
-Warn if comparing pointer parameter with nonnull attribute with NULL.
-
 Wnull-dereference
 Common Var(warn_null_dereference) Warning
 Warn if dereferencing a NULL pointer may lead to erroneous or undefined behavior.
